handler/enqueue_annotation.go

```diff
@@ -30,21 +32,56 @@ import (
+	// namespace-scoped owners and `<name>` for cluster-scoped owners.
 	NamespacedNameAnnotation = "operator-sdk/primary-resource"
-	// TypeAnnotation - annotation that will be used to verify that the primary resource is the primary resource to use.
+	// TypeAnnotation is an annotation whose value encodes the group and kind of a resource to reconcil when a
+	// resource containing this annotation changes. Valid values are of the form `<Kind>` for resource in the
+	// core group, and `<Kind>.<group>` for all other resources.
 	TypeAnnotation = "operator-sdk/primary-resource-type"
 )
 
-// EnqueueRequestForAnnotation enqueues Requests based on the presence of an annotation that contains the
-// namespaced name of the primary resource.
+// EnqueueRequestForAnnotation enqueues Request containing the Name and Namespace specified in the
+// annotations of the object that is the source of the Event. The source of the event triggers reconciliation
+// of the parent resource which is identified by annotations. `NamespacedNameAnnotation` and `TypeAnnotation` together uniquely
+// identify an owner resource to reconcile.
+// handler.EnqueueRequestForAnnotation can be used to trigger reconciliation of resources which are cross-referenced.
+// By that we mean that a namespace-scoped dependent can trigger reconciliation of an owner which is not cluster-scoped, and
+// a cluster-scoped dependent can trigger the reconciliation of a namespace(scoped)-owner.
+
+// As an example, consider the case where we would like to watch clusterRole resource based on which we reconcile ReplicaSets which are
+// namespace scoped. With native owner references, this would not be possible since the cluster-scoped dependent (ClusterRole) is trying to
+// specify a namespace-scoped owner (ReplicaSet). Whereas in case of annotations-based handlers, we could implement the following:
+
+// ...
+// if err := c.Watch(&source.Kind{
+//	  // Watch cluster roles
+//	  Type: &rbacv1.ClusterRole{}},
 //
-// The primary usecase for this, is to have a controller enqueue requests for the following scenarios
-// 1. namespaced primary object and dependent cluster scoped resource
-// 2. cluster scoped primary object.
-// 3. namespaced primary object and dependent namespaced scoped but in a different namespace object.
+//	  // Enqueue ReplicaSet reconcile requests using the
+//	  // namespacedName annotation value in the request.
+//	  &handler.EnqueueRequestForAnnotation{schema.GroupKind{Group:"ReplicaSet", Kind:"apps"}}); err != nil {
+//	      entryLog.Error(err, "unable to watch ClusterRole")
+//	      os.Exit(1)
+//    }
+// }
+// ...
+
+// With this watch, the ReplicaSet reconciler would receive a request to reconcile "my-namespace/my-replicaset" based
+// on a change to clusterRole that has the following annotations:
+//
+// annotations:
+//   	operator-sdk/primary-resource:"my-namespace/my-replicaset"
+//    	operator-sdk/primary-resource-type:"ReplicaSet.apps"
```

Comment:
```suggestion
//    	operator-sdk/primary-resource-type:"apps.ReplicaSet"
```

Reply:
This was actually correct before: https://github.com/kubernetes/apimachinery/blob/3f0ac6a8e1122b68c04c77e8947563cc8a8896bc/pkg/runtime/schema/group_version.go#L137